primitives/src/merkle_tree/internal.rs

```diff
@@ -655,6 +655,123 @@ where
             MerkleNode::ForgettenSubtree { .. } => Ok((0, LookupResult::NotInMemory)),
         }
     }
+
+    /// Persistently update the element at the given index.
+    /// * `returns` - `Err()` if any error happens internally. `Ok(delta,
+    ///   result)`, `delta` represents the changes to the overall number of
+    ///   leaves of the tree, `result` contains the original lookup information
+    ///   at the given location.
+    pub(crate) fn persistent_update_with_internal<H, Arity, F>(
+        &self,
+        height: usize,
+        pos: impl Borrow<I>,
+        traversal_path: &[usize],
+        f: F,
+    ) -> Result<(Box<Self>, i64), PrimitivesError>
+    where
+        H: DigestAlgorithm<E, I, T>,
+        Arity: Unsigned,
+        F: FnOnce(Option<&E>) -> Option<E>,
+    {
+        let pos = pos.borrow();
+        match self {
+            MerkleNode::Leaf {
+                elem: node_elem,
+                value: _,
+                pos,
+            } => match f(Some(node_elem)) {
+                Some(elem) => Ok((
+                    Box::new(MerkleNode::Leaf {
+                        value: H::digest_leaf(pos, &elem)?,
+                        pos: pos.clone(),
+                        elem,
+                    }),
+                    0i64,
+                )),
+                None => Ok((Box::new(MerkleNode::Empty), -1i64)),
+            },
+            MerkleNode::Branch { value, children } => {
+                let branch = traversal_path[height - 1];
+                let result = children[branch].persistent_update_with_internal::<H, Arity, _>(
+                    height - 1,
+                    pos,
+                    traversal_path,
+                    f,
+                )?;
+                let mut children = children.clone();
```

Comment:
I think this ends up doing a deep clone of the whole subtree, since each child is a `Box`. Don't we need to make the children `Arc` or something so we can shallow clone the children that aren't being updated?

Reply:
Then should we make everything `Arc`?
Also I learnt before that there are certain differences between `std::sync::Arc` and `async_std::sync::Arc`.

Reply:
Oh no way I thought those were just type aliases lol. For this I don't know that it matters too much which one we use (it's only going to be used internally anyways, not part of the public API).

I think we need to make everything `Arc`, but as I was saying in Zulip, this makes it hard to implement a regular mutable update. So I think we have two choices:
* Make everything `Arc`, and implement the regular mutable update by doing a non-mutable update and then, at the very end, setting the root with `*self = new_root`. This will have slightly worse performance but still O(log(n)) I think
* Have a completely separate node type for immutable and mutable updates

Personally I would lean towards the former as it means we have a lot less code to maintain (and mutable updates are a thin wrapper around immutable updates which is kind of nice). But I don't know how much we care about the performance of mutable updates.

Probably nothing we have now is highly sensitive to the performance of this, and because the mutable and immutable updates are in separate traits, we could always implement a specialized version of the tree that only supports mutable updates later on, if it became important

Reply:
Persistent data structure is a common name in competitive programming community. Anyway I change it to `NonDestructive`.

I've made everything `Arc`. Check the newest changes.

Reply:
Oh CI fails for wasm.

@alxiong Is there a wasm-compatible `Arc`?

Reply:
> Persistent data structure is a common name in competitive programming community. Anyway I change it to NonDestructive.

You learn something new every day! Thanks for introducing me to the terminology, and feel free to name it whatever you want if you like that better